File: packages/js/core/src/models/swin.ts
import { Tensor } from '../ir/tensor.js';
import { Gelu, Gemm, LayerNormalization, MultiHeadAttention } from '../primitives.js';
import { PatchEmbed } from './vit.js';

function getParam(name: string, shape: number[], dtype: any = 'float32'): Tensor {
  return new Tensor(name, shape, dtype, false, false, new Float32Array());
}

function recordOp(opType: string, inputs: Tensor[], attr?: any): Tensor {
  const dtype = inputs[0]?.dtype ?? 'float32';
  return new Tensor(`${opType}_out`, [], dtype, false, false, new Float32Array());
}

export class WindowAttention {
  public prefix: string;
  public dim: number;
  public windowSize: [number, number];
  public numHeads: number;
  public attn: MultiHeadAttention;
  public proj: Gemm;

  constructor(
    dim: number,
    windowSize: [number, number],
    numHeads: number,
    qkvBias: boolean = true,
    prefix: string = '',
  ) {
    this.prefix = prefix;
    this.dim = dim;
    this.windowSize = windowSize;
    this.numHeads = numHeads;
    this.attn = new MultiHeadAttention(numHeads, qkvBias, true);
    this.proj = new Gemm(1.0, 1.0, 0, 1);
  }

  call(x: Tensor): Tensor {
    x = this.attn.call(x, x, x);
    x = this.proj.call(
      x,
      getParam(`${this.prefix}.proj.weight`, [this.dim, this.dim]),
      getParam(`${this.prefix}.proj.bias`, [this.dim]),
    );
    return x;
  }
}

export class SwinTransformerBlock {
  public prefix: string;
  public dim: number;
  public inputResolution: [number, number];
  public numHeads: number;
  public windowSize: number;
  public shiftSize: number;
  public mlpRatio: number;

  public norm1: LayerNormalization;
  public attn: WindowAttention;
  public norm2: LayerNormalization;
  public mlpFc1: Gemm;
  public act: Gelu;
  public mlpFc2: Gemm;

  constructor(
    dim: number,
    inputResolution: [number, number],
    numHeads: number,
    windowSize: number = 7,
    shiftSize: number = 0,
    mlpRatio: number = 4.0,
    qkvBias: boolean = true,
    prefix: string = '',
  ) {
    this.prefix = prefix;
    this.dim = dim;
    this.inputResolution = inputResolution;
    this.numHeads = numHeads;
    this.windowSize = windowSize;
    this.shiftSize = shiftSize;
    this.mlpRatio = mlpRatio;

    if (Math.min(...this.inputResolution) <= this.windowSize) {
      this.shiftSize = 0;
      this.windowSize = Math.min(...this.inputResolution);
    }

    this.norm1 = new LayerNormalization([dim]);
    this.attn = new WindowAttention(
      dim,
      [this.windowSize, this.windowSize],
      numHeads,
      qkvBias,
      `${prefix}.attn`,
    );
    this.norm2 = new LayerNormalization([dim]);
    this.mlpFc1 = new Gemm(1.0, 1.0, 0, 1);
    this.act = new Gelu();
    this.mlpFc2 = new Gemm(1.0, 1.0, 0, 1);
  }

  call(x: Tensor): Tensor {
    let identity = x;
    x = this.norm1.call(
      x,
      getParam(`${this.prefix}.norm1.weight`, [this.dim]),
      getParam(`${this.prefix}.norm1.bias`, [this.dim]),
    );

    if (this.shiftSize > 0) {
      x = recordOp('Roll', [x], { shifts: [-this.shiftSize, -this.shiftSize], axes: [1, 2] });
    }

    x = this.attn.call(x);

    if (this.shiftSize > 0) {
      x = recordOp('Roll', [x], { shifts: [this.shiftSize, this.shiftSize], axes: [1, 2] });
    }

    x = recordOp('Add', [x, identity]);

    identity = x;
    x = this.norm2.call(
      x,
      getParam(`${this.prefix}.norm2.weight`, [this.dim]),
      getParam(`${this.prefix}.norm2.bias`, [this.dim]),
    );

    const mlpDim = Math.floor(this.dim * this.mlpRatio);
    x = this.mlpFc1.call(
      x,
      getParam(`${this.prefix}.mlp_fc1.weight`, [mlpDim, this.dim]),
      getParam(`${this.prefix}.mlp_fc1.bias`, [mlpDim]),
    );
    x = this.act.call(x);
    x = this.mlpFc2.call(
      x,
      getParam(`${this.prefix}.mlp_fc2.weight`, [this.dim, mlpDim]),
      getParam(`${this.prefix}.mlp_fc2.bias`, [this.dim]),
    );
    x = recordOp('Add', [x, identity]);

    return x;
  }
}

export class SwinTransformer {
  public numClasses: number;
  public embedDim: number;
  public patchEmbed: PatchEmbed;
  public layers: SwinTransformerBlock[];
  public norm: LayerNormalization;
  public head: Gemm;

  constructor(
    embedDim: number = 96,
    depths: number[] = [2, 2, 6, 2],
    numHeads: number[] = [3, 6, 12, 24],
    windowSize: number = 7,
    mlpRatio: number = 4.0,
    numClasses: number = 1000,
  ) {
    this.numClasses = numClasses;
    this.embedDim = embedDim;
    this.patchEmbed = new PatchEmbed(224, 4, 3, embedDim, 'patch_embed');

    this.layers = [];
    for (let iLayer = 0; iLayer < depths.length; iLayer++) {
      const dim = Math.floor(embedDim * Math.pow(2, iLayer));
      const inputResolution: [number, number] = [
        Math.floor(224 / (4 * Math.pow(2, iLayer))),
        Math.floor(224 / (4 * Math.pow(2, iLayer))),
      ];

      for (let i = 0; i < depths[iLayer]!; i++) {
        const shiftSize = i % 2 === 0 ? 0 : Math.floor(windowSize / 2);
        this.layers.push(
          new SwinTransformerBlock(
            dim,
            inputResolution,
            numHeads[iLayer]!,
            windowSize,
            shiftSize,
            mlpRatio,
            true,
            `layers.${iLayer}.blocks.${i}`,
          ),
        );
      }
    }

    const lastDim = Math.floor(embedDim * Math.pow(2, depths.length - 1));
    this.norm = new LayerNormalization([lastDim]);
    this.head = new Gemm(1.0, 1.0, 0, 1);
  }

  call(x: Tensor): Tensor {
    x = this.patchEmbed.call(x);

    for (const block of this.layers) {
      x = block.call(x);
    }

    x = recordOp('ReduceMean', [x], { axes: [1], keepdims: 0 });

    const lastDim = Math.floor(this.embedDim * Math.pow(2, 3));
    x = this.norm.call(x, getParam('norm.weight', [lastDim]), getParam('norm.bias', [lastDim]));
    x = this.head.call(
      x,
      getParam('head.weight', [this.numClasses, lastDim]),
      getParam('head.bias', [this.numClasses]),
    );

    return x;
  }
}

export function swinT(numClasses: number = 1000): SwinTransformer {
  return new SwinTransformer(96, [2, 2, 6, 2], [3, 6, 12, 24], 7, 4.0, numClasses);
}

File: packages/python/onnx9000-core/src/onnx9000/core/models/swin.py
"""Swin."""

from typing import Any

from onnx9000.core.ir import Tensor, Variable
from onnx9000.core.ops import add, concat, flatten, reshape, roll
from onnx9000.core.primitives import Gelu, Gemm, LayerNormalization, MultiHeadAttention


def get_param(name: str, shape: list[int], dtype: int = 1) -> Variable:  # noqa: D103
    """Get param."""
    return Variable(name=name, shape=shape, dtype=dtype)


class WindowAttention:
    """Window based multi-head self attention (W-MSA) module with relative position bias."""

    def __init__(  # noqa: D107
        self,
        dim: int,
        window_size: tuple[int, int],
        num_heads: int,
        qkv_bias: bool = True,
        prefix: str = "",
    ):
        """Init."""
        self.prefix = prefix
        self.dim = dim
        self.window_size = window_size
        self.num_heads = num_heads
        self.attn = MultiHeadAttention(num_heads=num_heads, qkv_bias=qkv_bias)
        self.proj = Gemm(trans_b=1)

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        # B_, N, C = x.shape
        # Just use standard attention. The shift is handled outside.
        # Adding relative positional encoding is usually a bias
        # For simplicity in AST builder, we just add the bias.

        # In actual swin, the rel_pos_bias is added inside attention (to attn_weight).
        # We can pass it as mask to MultiHeadAttention if supported, but let's just
        # assume our MHA can take an alibi-like bias or we skip it for the high-level AST.

        """Call."""
        x = self.attn(x, x, x)
        x = self.proj(
            x,
            get_param(f"{self.prefix}.proj.weight", [self.dim, self.dim]),
            get_param(f"{self.prefix}.proj.bias", [self.dim]),
        )
        return x


class SwinTransformerBlock:
    """Swin Transformer Block."""

    def __init__(  # noqa: D107
        self,
        dim: int,
        input_resolution: tuple[int, int],
        num_heads: int,
        window_size: int = 7,
        shift_size: int = 0,
        mlp_ratio: float = 4.0,
        qkv_bias: bool = True,
        prefix: str = "",
    ):
        """Init."""
        self.prefix = prefix
        self.dim = dim
        self.input_resolution = input_resolution
        self.num_heads = num_heads
        self.window_size = window_size
        self.shift_size = shift_size
        self.mlp_ratio = mlp_ratio

        if min(self.input_resolution) <= self.window_size:
            self.shift_size = 0
            self.window_size = min(self.input_resolution)

        self.norm1 = LayerNormalization((dim,))
        self.attn = WindowAttention(
            dim,
            window_size=(self.window_size, self.window_size),
            num_heads=num_heads,
            qkv_bias=qkv_bias,
            prefix=f"{prefix}.attn",
        )
        self.norm2 = LayerNormalization((dim,))
        self.mlp_fc1 = Gemm(trans_b=1)
        self.act = Gelu()
        self.mlp_fc2 = Gemm(trans_b=1)

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        """Call."""
        identity = x
        x = self.norm1(
            x,
            get_param(f"{self.prefix}.norm1.weight", [self.dim]),
            get_param(f"{self.prefix}.norm1.bias", [self.dim]),
        )

        if self.shift_size > 0:
            # shifted window
            x = roll(x, shifts=[-self.shift_size, -self.shift_size], axes=[1, 2])

        x = self.attn(x)

        if self.shift_size > 0:
            # reverse shift
            x = roll(x, shifts=[self.shift_size, self.shift_size], axes=[1, 2])

        x = add(x, identity)

        identity = x
        x = self.norm2(
            x,
            get_param(f"{self.prefix}.norm2.weight", [self.dim]),
            get_param(f"{self.prefix}.norm2.bias", [self.dim]),
        )
        x = self.mlp_fc1(
            x,
            get_param(f"{self.prefix}.mlp_fc1.weight", [int(self.dim * self.mlp_ratio), self.dim]),
            get_param(f"{self.prefix}.mlp_fc1.bias", [int(self.dim * self.mlp_ratio)]),
        )
        x = self.act(x)
        x = self.mlp_fc2(
            x,
            get_param(f"{self.prefix}.mlp_fc2.weight", [self.dim, int(self.dim * self.mlp_ratio)]),
            get_param(f"{self.prefix}.mlp_fc2.bias", [self.dim]),
        )
        x = add(x, identity)

        return x


class SwinTransformer:  # noqa: D101
    """Swin transformer."""

    def __init__(  # noqa: D107
        self,
        embed_dim: int = 96,
        depths: list[int] = None,
        num_heads: list[int] = None,
        window_size: int = 7,
        mlp_ratio: float = 4.0,
        num_classes: int = 1000,
    ):
        """Init."""
        if num_heads is None:
            num_heads = [3, 6, 12, 24]
        if depths is None:
            depths = [2, 2, 6, 2]
        self.num_classes = num_classes
        self.embed_dim = embed_dim
        # Patch embedding is basically a Conv2D with stride = patch_size
        from onnx9000.core.models.vit import PatchEmbed

        self.patch_embed = PatchEmbed(
            img_size=224, patch_size=4, in_chans=3, embed_dim=embed_dim, prefix="patch_embed"
        )

        # Build layers (blocks)
        self.layers = []
        for i_layer in range(len(depths)):
            dim = int(embed_dim * 2**i_layer)
            input_resolution = (224 // (4 * 2**i_layer), 224 // (4 * 2**i_layer))

            for i in range(depths[i_layer]):
                shift_size = 0 if (i % 2 == 0) else window_size // 2
                self.layers.append(
                    SwinTransformerBlock(
                        dim=dim,
                        input_resolution=input_resolution,
                        num_heads=num_heads[i_layer],
                        window_size=window_size,
                        shift_size=shift_size,
                        mlp_ratio=mlp_ratio,
                        prefix=f"layers.{i_layer}.blocks.{i}",
                    )
                )

            # Patch Merging at the end of each layer except the last
            # Usually implemented with a linear layer that concatenates adjacent patches
            # Omitting the explicit patch merging for brevity in AST, or we can use a linear layer
            if i_layer < len(depths) - 1:
                # We would add PatchMerging here
                continue

        self.norm = LayerNormalization((int(embed_dim * 2 ** (len(depths) - 1)),))
        self.head = Gemm(trans_b=1)

    def __call__(self, x: Tensor) -> Tensor:  # noqa: D102
        """Call."""
        x = self.patch_embed(x)

        for block in self.layers:
            x = block(x)

        # Global average pool over sequence length
        from onnx9000.core.ops import reduce_mean

        x = reduce_mean(x, axes=[1], keepdims=False)

        last_dim = int(self.embed_dim * 2**3)
        x = self.norm(x, get_param("norm.weight", [last_dim]), get_param("norm.bias", [last_dim]))
        x = self.head(
            x,
            get_param("head.weight", [self.num_classes, last_dim]),
            get_param("head.bias", [self.num_classes]),
        )

        return x


def swin_t(**kwargs: Any) -> SwinTransformer:  # noqa: D103
    """Swin t."""
    return SwinTransformer(**kwargs)
